### ISSUE TITLE
[ feature request]

### ISSUE THREAD

**Issue by vince62s** [google/sentencepiece#142]
As of now, it is possible to specify user defined symbol to by pass some sequences.
It would be great if we could pass a "pattern" for these symbol especially when we have plenty of placeholders.
for instance pattern could be '(((*)))' and the * would be any kind of string.
cheers.

**Comment by taku910** [comment]
Let me clarify this request.

For example when we pass --user_defined_symbols="<*>",  all strings that match "<*>" in the corpus are treated one unique symbol?

That would be technically possible, but need to scan the corpus twice to enumerate the symbols.

**Comment by vince62s** [comment]
no it's not the request.
My understanding is that user_defined_symbol can be used for placeholders that need NOT tobe tokenized.
instead of a defined list, I would like a pattern.
then I could use multiple symbols with the same pattern, none of these would be tokenized and kept intact.
clearer ?

**Comment by taku910** [comment]
Right, but user_defined_symbols have corresponding vocab_id and we have to enumerate them prior to model training.

It would be possible to specify some patterns not to tokenize, but this means that these dynamically extracted symbols will not always have vocab id and be treated as unknown symbol.

**Comment by vince62s** [comment]
I don't think we need a vocab_id for these.

**Comment by taku910** [comment]
All tokens must have an id. 
SentencePiece is not a segmenter, but an encoder that convert the text into a one-hot vector sequence.

**Comment by vince62s** [comment]
ok then what about:
at training using a unque vocab_id (same as for the 3 special tokens)
at encoding using a mechanism to "keep" the special token from source (same a replace_unk mechanism used in nmt inference)
not sure if I'm clear and if it may work.

**Comment by taku910** [comment]
This sounds basically the same as handling them as unknown token (id=0). This would technically be possible, but it is not consistent with the original motivation of subword units, which attempts to reduce unknown words.

Please let me know how are you going to use this pattern based specification? What the application looks like?   I'm not quite sure the situation where increasing unknown symbols works at least in neural-based system.

**Comment by vince62s** [comment]
use case is placeholders to handle tags, but with possibly several tenth (maybe hundreds) of tags.
I don't need the placeholders/tags to be in the SP vocab, but they need to be preserved intact at tokenization. My data preprocessing (after tokenization) will handle them as tokens and include them in the NMT vocab.
let me know if you need more clarification.
